include/core/parseutil.h

```diff
@@ -39,42 +39,41 @@ class Token {
-    QString readCIncbin(QString text, QString label);
-    QMap<QString, int> readCDefines(QString filename, QStringList prefixes, QMap<QString, int> = QMap<QString, int>());
-    void readCDefinesSorted(QString, QStringList, QStringList*, QMap<QString, int> = QMap<QString, int>());
-    QList<QStringList> getLabelMacros(const QList<QStringList> &, const QString &);
-    QStringList getLabelValues(const QList<QStringList> &, const QString &);
-    bool tryParseJsonFile(QJsonDocument *out, QString filepath);
-    bool ensureFieldsExist(QJsonObject obj, QList<QString> fields);
+    int evaluateDefine(const QString&, const QMap<QString, int>&);
+    QStringList readCArray(const QString &text, const QString &label);
+    QMap<QString, QString> readNamedIndexCArray(const QString &text, const QString &label);
+    QString readCIncbin(const QString &text, const QString &label);
+    QMap<QString, int> readCDefines(const QString &filename, const QStringList &prefixes, QMap<QString, int> = { });
+    QStringList readCDefinesSorted(const QString&, const QStringList&, const QMap<QString, int>& = { });
+    QList<QStringList> getLabelMacros(const QList<QStringList>&, const QString&);
+    QStringList getLabelValues(const QList<QStringList>&, const QString&);
+    bool tryParseJsonFile(QJsonDocument *out, const QString &filepath);
+    bool ensureFieldsExist(const QJsonObject &obj, const QList<QString> &fields);
 
     // Returns the 1-indexed line number for the definition of scriptLabel in the scripts file at filePath.
     // Returns 0 if a definition for scriptLabel cannot be found.
     static int getScriptLineNumber(const QString &filePath, const QString &scriptLabel);
     static int getRawScriptLineNumber(QString text, const QString &scriptLabel);
     static int getPoryScriptLineNumber(QString text, const QString &scriptLabel);
-    static QString &removeStringLiterals(QString &text);
-    static QString &removeLineComments(QString &text, const QString &commentSymbol);
-    static QString &removeLineComments(QString &text, const QStringList &commentSymbols);
+    static QString removeStringLiterals(QString text);
+    static QString removeLineComments(QString text, const QString &commentSymbol);
+    static QString removeLineComments(QString text, const QStringList &commentSymbols);
 
     static QStringList splitShellCommand(QStringView command);
 
 private:
     QString root;
     QString text;
     QString file;
-    QList<Token> tokenizeExpression(QString expression, QMap<QString, int>* knownIdentifiers);
-    QList<Token> generatePostfix(QList<Token> tokens);
-    int evaluatePostfix(QList<Token> postfix);
-    void error(QString message, QString expression);
+    QList<Token> tokenizeExpression(QString expression, const QMap<QString, int> &knownIdentifiers);
+    QList<Token> generatePostfix(const QList<Token> &tokens);
+    int evaluatePostfix(const QList<Token> &postfix);
```

Comment:
What's the reason you changed so many function signatures to use `const &`?

Reply:
The rest looks good to me.  I'm just not sure why a lot of these are using `const &` now.  When the object is copied, it probably just the reference container, and not the underlying map or list items themselves.

Reply:
It's true that copying `Qt`'s shared-data classes is less expensive than copying the corresponding `std` class for instance, but it's generally a good idea not to copy if you don't need to. There is *some* overhead to copying `Qt`'s data classes, even if it is minimal. I think maybe the best argument for passing by `const &` in these cases is that all of `Qt`'s own functions do so for container classes. My general rule of thumb is to only pass containers by value if the function actually needs a copy.